xc/device.ml

```diff
@@ -2170,32 +2160,30 @@ module Backend = struct
         )
 
       let suspend (task: Xenops_task.task_handle) ~xs ~qemu_domid domid =
-        let open Qmp in
-        let c = Qmp_protocol.connect (qmp_libxl_path domid) in
+        let as_msg cmd = Qmp.(Success(Some __LOC__, cmd)) in
+        let perms      = [ Unix.O_WRONLY; Unix.O_CREAT ] in
+        let save_file  = sprintf qemu_save_path domid in
+        let save_fd    = Unix.openfile save_file perms 0o660 in
         finally
           (fun () ->
-             let fd_of_c = Qmp_protocol.to_fd c in
-             let save_file = sprintf qemu_save_path domid in
-             let fd_of_save_file = Unix.openfile save_file [ Unix.O_WRONLY; Unix.O_CREAT ] 0o660 in
-             finally
-               (fun () -> ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_save_file))
-               (fun () -> Unix.close fd_of_save_file);
-
-             let qmp_cmd = Command (None, Add_fd None) in
-             Qmp_protocol.negotiate c;
-             Qmp_protocol.write c qmp_cmd;
-             let fd_info = match Qmp_protocol.read c with
-               | Success (None, Fd_info x) -> x
-               | _ -> raise (Internal_error (sprintf "Get unexpected result after sending Qmp message: %s" (string_of_message qmp_cmd)))
-             in
-             finally
-               (fun () ->
-                  let path = sprintf "/dev/fdset/%d" fd_info.fdset_id in
-                  Qmp_protocol.write c (Command (None, Stop));
-                  Qmp_protocol.write c (Command (None, Xen_save_devices_state path)))
-               (fun () ->
-                  Qmp_protocol.write c (Command (None, Remove_fd fd_info.fdset_id))))
-          (fun () -> Qmp_protocol.close c)
+            let fd = qmp_send_cmd ~send_fd:save_fd domid Qmp.(Add_fd None)
+              |> function
+              | Qmp.(Fd_info fd) -> fd
+              | other ->
+                  raise (Internal_error (sprintf
+                    "Unexpected result for QMP command: %s"
+                    Qmp.(other |> as_msg |> string_of_message)))
```

Comment:
as_msg is a Qmp.Success. Shouldn't it be a Qmp.Error?

Reply:
No. If it were `Error`, an exception would have been raised by `qmp_send_cmd`. This is the case that we have `Success` but it carries unexpected data. However, this is a big ugly we are missing a function to turn a result (and not a message) into a string in the Qmp library.